Support `FILTER` in over clause

lovasoa opened the pull request:
support FILTER clause in window functions

fixes #1006

lovasoa commented:
@alamb , I'd love to get your feedback on this :slightly_smiling_face: 

As before, I need this for [SQLPage](https://github.com/lovasoa/SQLpage)

alamb commented on src/ast/mod.rs:
```diff
@@ -991,8 +991,11 @@ impl Display for WindowType {
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "visitor", derive(Visit, VisitMut))]
 pub struct WindowSpec {
+    // OVER (PARTITION BY ...)
```
If we use `///` then these will show up in documentation as well:

```suggestion
    /// `OVER (PARTITION BY ...)`
```

alamb commented on src/ast/mod.rs:
```diff
@@ -991,8 +991,11 @@ impl Display for WindowType {
 pub struct WindowSpec {
+    // OVER (PARTITION BY ...)
     pub partition_by: Vec<Expr>,
+    // OVER (ORDER BY ...)
```
```suggestion
    /// `OVER (ORDER BY ...)`
```

alamb commented on src/ast/mod.rs:
```diff
@@ -991,8 +991,11 @@ impl Display for WindowType {
     pub partition_by: Vec<Expr>,
+    // OVER (ORDER BY ...)
     pub order_by: Vec<OrderByExpr>,
+    // OVER (window frame)
```
```suggestion
    /// `OVER (window frame)`
```

alamb commented on tests/sqlparser_sqlite.rs:
```diff
@@ -290,6 +290,31 @@ fn parse_create_table_with_strict() {
 
+#[test]
+fn parse_window_function_with_filter() {
+    let sql = "SELECT max(x) FILTER (WHERE y) OVER () FROM t";
```
I wonder if you  can add another test with for a window function (aka that has an `OVER` clause) but that is not an aggregate. For example:

```sql
SELECT row_number() FILTER (WHERE y) OVER () FROM t
```

I tried it in postgres and found the syntax is parsed but the functionality is not implemented

```
ERROR:  FILTER is not implemented for non-aggregate window functions
LINE 1: SELECT row_number() FILTER (WHERE column1 > 1) OVER () FROM ...
```

lovasoa replied:
done in https://github.com/sqlparser-rs/sqlparser-rs/pull/1007/commits/dd51436086f0123abe7a31bb24d5515b936a091d

alamb commented on src/ast/mod.rs:
```diff
@@ -3650,6 +3653,8 @@ impl fmt::Display for CloseCursor {
     pub name: ObjectName,
     pub args: Vec<FunctionArg>,
+    // e.g. `x > 5` in `COUNT(x) FILTER (WHERE x > 5)`
+    pub filter: Option<Box<Expr>>,
```
The current way to represent aggregates with filters (in the select clause) is https://docs.rs/sqlparser/latest/sqlparser/ast/enum.Expr.html#variant.AggregateExpressionWithFilter

The idea of adding the `filter` directly to the Function I think is more elegant and general

What do you think about (in some other PR) removing `AggregateExpressionWithFilter` and just using this more general formulation?

lovasoa replied:
Yes, that is a good idea :+1:

lovasoa replied:
But we would have to make sure that supporting things like `(x + y) filter (where z)` was actually accidental

alamb commented on src/dialect/sqlite.rs:
```diff
@@ -35,6 +35,10 @@ impl Dialect for SQLiteDialect {
             || ('\u{007f}'..='\u{ffff}').contains(&ch)
     }
 
+    fn supports_filter_during_aggregation(&self) -> bool {
```
This PR supports filtering for any function with an `OVER` clause, correct? Thus, this name is no longer quite right I don't think (again maybe this is something we change in a follow on PR) 🤔

lovasoa replied:
Yes, let's look at that in a separate pr. This may need some investigation to make sure we do not remove useful functionality

alamb reviewed (approved):
Thanks @lovasoa -- I have a few suggestions and comments, but I also think we can merge this PR as is

lovasoa commented:
@alamb : thanks for the review, I addressed your comments. We can see about removing AggregateExpressionWithFilter and supports_filter_during_aggregation, but I'd be careful because it will prevent parsing some statements that parsed before.

lovasoa commented on src/ast/mod.rs:
```diff
@@ -3650,6 +3653,8 @@ impl fmt::Display for CloseCursor {
 pub struct Function {
     pub name: ObjectName,
     pub args: Vec<FunctionArg>,
+    // e.g. `x > 5` in `COUNT(x) FILTER (WHERE x > 5)`
```
```suggestion
    /// e.g. `x > 5` in `COUNT(x) FILTER (WHERE x > 5)`
```

alamb reviewed (approved):
Thanks @lovasoa  -- I still thing something is confused here as I don't think SQL supports things like `select foo FILTER(where x > 5) FROM t` which I think this PR does allow. 

However I think this is a reasonable improvement

It would be great to sort out the naming of `Dialect::supports_filter_during_aggregation` as that now also controls `filter during any function`
